### ISSUE TITLE
Popups in firefox >= 92.0b2 are broken with scale != 1

### ISSUE THREAD

**Issue by ojab** [swaywm/sway#6426]
- **Sway Version:**
  - sway version 1.6-d0fe721f (Aug  9 2021, branch 'master')
  - wlroots 88f65db87f8c6cc1f5598ef1b782a7cef6303dd2 (latest master at the moment)
  
- **Debug Log:**
  - https://gist.github.com/b5378de60dcf1e60d083927446298ca5
  - `exec firefox` in a sway config, then clicked menu button, then pocket button

- **Description:**
Since [firefox bugs are allowed now](https://github.com/swaywm/sway/commit/db0d63313dec8a3036d06ef9a69271bac6bfa531), let's report https://bugzilla.mozilla.org/show_bug.cgi?id=1722767 here.
Popups (both menus and right-click) in the current firefox nightly are not visible in sway.

**Comment by ojab** [comment]
https://gist.github.com/ojab/150cd7cc1f44af235c6a70aba8d3e3b3 sway+firefox log with ` MOZ_LOG="WidgetScreen:5 WidgetPopup:5"`, maybe it's more clear what's happening this way.

**Comment by wkoomson** [comment]
This might not be relevant since I am on Firefox v90.0.2 (not nightly) and sway v1.5, but  I experience the same behavior, i.e., clicking on a toolbar item (such as Pocket or Bitwarden) and the popup window never shows up. Also, sometimes right click menus do not show.

Interestingly enough, I only see this behavior on Firefox windows on external displays. I do not see this on Firefow windows on my main laptop display (i.e. eDP-1).

A random workaround that seems to work: after clicking on the toolbar item, toggle the floating mode of the Firefox window (e.g., if tiled, switch to floating and if floating switch to tiled), and then the popup window will appear. It seems like the popup window is created when you first click the toolbar item, but for some reason is not showing up until you toggle.

**Comment by ojab** [comment]
Looks like you're having https://github.com/swaywm/sway/issues/6147, this is different (bisected till a single Firefox commit and reproducible with `!= 1` scale on the single monitor).

**Comment by stransky** [comment]
From some reason we don't get callback from xdg_positioner for the popup so it's not shown then.

**Comment by alanorth** [comment]
As of 92.0b2 they are broken in Firefox Beta / Developer Edition as well. On both Sway 1.6.1 and master.

**Comment by ojab** [comment]
This particular bug cannot be reproduced in Firefox 91.0 since change with regression is not in Firefox 91.0 (Firefox 92.0 would include it).

**Comment by emersion** [comment]
Can you collect a `WAYLAND_DEBUG=1` log of Firefox when reproducing this bug?

**Comment by ojab** [comment]
https://gist.github.com/3cbd8ccdb09947c0c9279578f8838076
Will do along with `sway -d` if needed.

**Comment by ojab** [comment]
hmm, tried to reproduce it with latest nightly in sway master from the first message and it's not reproducible anymore, although still no popups in sway-1.6 (log in the message above).
So something is changed in firefox in the last few days and it started to work fine in sway master.

**Comment by luispabon** [comment]
Do you have any scaled outputs? If all your outputs are scale=1 popups do work.

Broken for me on FF 93.0~a1~hg20210810r588305-0ubuntu0.21.04.1~umd1

**Comment by ojab** [comment]
Yes, single output with scale 1.5. Do you use sway from master?

**Comment by luispabon** [comment]
No, I'm pegged to 1.6.1 at the moment. I'll check on wlroots/sway master later today after work

**Comment by luispabon** [comment]
I've upgraded now to sway & wlroots master and the problem remains.

**Comment by RPigott** [comment]
@stransky sway doesn't respond to the callback because the surface is destroyed.

Here is logging typical of a failed popup on scale != 1:

<details>
<summary>
`MOZ_LOG="WidgetWayland:5 WidgetScreen:5 WidgetPopup:5" WAYLAND_DEBUG=1 firefox --gdk-debug=events
</summary>

```
[Parent 43945: Main Thread]: D/WidgetWayland moz_container_wayland_map_event [7facfc555f50]
[Parent 43945: Main Thread]: D/WidgetWayland moz_container_wayland_surface_create_locked [7facfc555f50]
[Parent 43945: Main Thread]: D/WidgetWayland     gtk wl_surface 7facf3f6f600 ID 120
[882711.198]  -> wl_compositor@35.create_surface(new id wl_surface@123)
[882711.222]  -> wl_subcompositor@36.get_subsurface(new id wl_subsurface@124, wl_surface@123, wl_surface@120)
[882711.235]  -> wl_subsurface@124.set_desync()
[882711.247]  -> wl_surface@120.frame(new id wl_callback@125)
[Parent 43945: Main Thread]: D/WidgetWayland     created frame callback ID 125
[882711.264]  -> wl_surface@123.commit()
[Parent 43945: Main Thread]: D/WidgetWayland     created surface 7facf2c84330 ID 123
[Parent 43945: Main Thread]: D/WidgetWayland moz_container_wayland_set_scale_factor_locked [7facfc555f50] scale 2
[882711.335]  -> wl_surface@123.set_buffer_scale(2)
[882711.346]  -> wl_compositor@35.create_region(new id wl_region@126)
[882711.356]  -> wl_surface@123.set_input_region(wl_region@126)
[882711.365]  -> wl_region@126.destroy()
[Parent 43945: Main Thread]: D/WidgetWayland moz_container_wayland_invalidate [7facfc555f50]
[Parent 43945: Main Thread]: D/WidgetPopup nsWindow::OnWindowStateEvent [7facf3fa6800] for 7facf3fa5660 changed 0x1 new_window_state 0x80
[Parent 43945: Main Thread]: D/WidgetPopup 	early return because no interesting bits changed
[Parent 43945: Main Thread]: D/WidgetPopup nsWindow::OnWindowStateEvent [7facf3fa6800] for 7facfc555f50 changed 0x1 new_window_state 0x80
[Parent 43945: Main Thread]: D/WidgetPopup 	quick return because IS_MOZ_CONTAINER(aWidget) is true
[882711.796] wl_surface@120.enter(wl_output@31)
Gdk-Message: 12:53:29.589: surface enter, window 0x7facf6917640 output 0x7fad19cd1150
[882711.885] wl_surface@120.enter(wl_output@17)
Gdk-Message: 12:53:29.589: surface enter, window 0x7facf6917640 output 0x7fad2c6f5830
[882711.960]  -> xdg_popup@118.destroy()
[882711.972]  -> xdg_surface@101.destroy()
[882711.989]  -> wl_surface@120.destroy()
[...]
[882713.773] wl_display@1.delete_id(125) # callback is destroyed b/c the surface for this callback has been destroyed
[...]
[Parent 43945: Renderer]: D/WidgetWayland WindowSurfaceWaylandMB::Commit [7facf2c94c10] frame queued: can't lock wl_surface
```
</details>

In this case the popup was atually killed before the frame event was even commited (is this a typo? FF immediately commits the subsurface & flushes the display after requesting a callback on the parent). At the time firefox complains about not being able to obtain the container lock, the frame handler that's supposed to unlock it is waiting for a callback that doesn't exist.

The frame is requested here: https://searchfox.org/mozilla-central/source/widget/gtk/MozContainerWayland.cpp#527

The surface is destroyed by gtk in `gdk_wayland_window_hide_surface` here: https://gitlab.gnome.org/GNOME/gtk/-/blob/d4e2d05cd9518ba04d6fbe1cbcec27142788ac95/gdk/wayland/gdkwindow-wayland.c#L3396

The output.enter event triggers it in gtk via `surface_enter -> window_update_scale -> gdk_wayland_window_maybe_configure -> gdk_window_hide`: https://gitlab.gnome.org/GNOME/gtk/-/blob/d4e2d05cd9518ba04d6fbe1cbcec27142788ac95/gdk/wayland/gdkwindow-wayland.c#L1227

**Comment by RPigott** [comment]
Here's the full log from the excerpt: https://gist.github.com/RPigott/09580624cb1825d2909b5a5eb411daf1

I added a log message in `gdk_wayland_window_hide_surface`.

To capture this log, I ran `MOZ_LOG="WidgetWayland:5 WidgetScreen:5 WidgetPopup:5" WAYLAND_DEBUG=1 firefox --gdk-debug=events --private-window 'http://example.com'`, then right clicked to open and close a popup context menu on my scale=1 display. I then set my display scale to 1.5 and attempted to open a context menu again, but it does not appear. The excerpt is from this second popup.

**Comment by yourealwaysbe** [comment]
I have either this issue or one very close on Firefox 91.0 and Sway 1.6.1-1 (Arch Linux).

No scaling on Sway, but in Firefox with `layout.css.devPixelsPerPx` set to 1.4 menus from the navigation bar (addons &c.) are invisible but flicker as i move the mouse over them. Setting the property to 1.39 instead leads to stable addon menus, but the bookmarks menu remaining unusable.

**Comment by szbergeron** [comment]
Same issue is reproducible in Wayfire on top of wlroots master for me

**Comment by GreyXor** [comment]
Same problem for me

Firefox nightly (Built from https://hg.mozilla.org/mozilla-central/rev/405034da0bdf302611bdd6d6a61ef55bd7be4933)
Sway version 1.6-d0c5c5a6 (Aug 17 2021, branch 'master')
wlroots version : 59b292b69165d2dee0b2398ed2d0b38dea98d90f

**Comment by Jamedjo** [comment]
@RPigott Thanks for the detailed investigation! Useful to see how you went about investigating this and will really help to have this fixed.

It sounds like this partially relates to the fix in [bugzilla.gnome #772505](https://bugzilla.gnome.org/show_bug.cgi?id=772505)? Should we open an issue at https://gitlab.gnome.org/GNOME/gtk? Or is this more on the Firefox side still?

[GNOME/gtk@83b54 - Fix a race condition with xdg_popup resize](https://gitlab.gnome.org/GNOME/gtk/-/commit/83b54bab5734f17d01633c1b5a0ce3e4896d2ee0) adds the show/hide behaviour:

> When resizing an xdg_popup immediately after the initial mapping, there is a race condition between the client and the compositor which is processing the initial size given by the xdg_positioner, leading to the xdg_popup to be eventually of the wrong size.
>
> Only way to make sure the size is correct in that case is to hide and show the window again. Considering this occurs before the initial configure is processed, it should not be noticeable.

-----

Some comments from [GTK bug #772505](https://bugzilla.gnome.org/show_bug.cgi?id=772505) that necessitated the hiding and re-showing:

> I am somehow tempted to think it's an application issue, I don't think it's wise or advisable for the application to update the menu content in a timeout callback, is it?

> I agree, though there are proper use-cases where you've no other option than run the checks asynchronously. 

----

> The surface is destroyed by gtk in gdk_wayland_window_hide_surface here

It looks like a different approach was tried in [GNOME/gtk@0f2e19 Push NULL buffer when hiding a GdkWindow](https://gitlab.gnome.org/GNOME/gtk/-/commit/0f2e19c01059d151d606e7006d95dc7f60c222fc), but then reverted in favour of [GNOME/gtk@4ebf98](https://gitlab.gnome.org/GNOME/gtk/-/commit/4ebf98f440cda4b5e5b6fc72c9623acbeca1f81b).

> This is how windows are meant to be hidden as per the wayland protocol, there's no need to destroy the xdg_surface and other
interfaces.

----

It also looks like there was originally a boolean to decide if the surface should be destroyed, but that was removed in [GNOME/gtk@87e217 Destroy the wl_surface too when hiding a window ](https://gitlab.gnome.org/GNOME/gtk/-/commit/87e2a7d4b23e633d0f2263d6ebd205c113eede0f):

> wl_surfaces can't switch roles, so destroying the xdg_surface but not the wl_surface means that we could get an error when trying to re-map the surface.
> 
> We could fix this by not destroying the xdg resource and only do it at finalization time, but it's just as easy to just create a new wl_surface.

**Comment by RPigott** [comment]
@Jamedjo

> Should we open an issue at https://gitlab.gnome.org/GNOME/gtk? Or is this more on the Firefox side still?

It does seem like unusual behavior from gtk to me, so it would be fair to call it a gtk bug, especially since there doesn't seem to be any kind of signal from gtk that the surface is destroyed even though it is possible for firefox to get a reference to that surface with `gdk_wayland_window_get_wl_surface` as it does. However firefox is using gtk3 and may need a workaround even if the behavior in gtk4 is different.

> It looks like a different approach was tried in GNOME/gtk@0f2e19 Push NULL buffer when hiding a GdkWindow, but then reverted in favour of GNOME/gtk@4ebf98.

FWIW, the justification given by daniels to scrap the NULL buffer approach is really outdated. It was only [disallowed on mutter with the obsolete xdg_shell v6 protocol](https://gitlab.gnome.org/GNOME/mutter/-/blob/main/src/wayland/meta-wayland-legacy-xdg-shell.c#L1384) and tbh I don't understand why. wlroots and mutter both allow this with xdg_shell stable today. It would probably work.

**Comment by ian-s-mcb** [comment]
I'm in the same boat as @yourealwaysbe : flickering menus caused by a fractional scale, not with a Sway output, but with the Firefox `about:config` setting `layout.css.devPixelsPerPx`.

I created this Firefox bug report in case our issue and #6426 have a different root cause.

https://bugzilla.mozilla.org/show_bug.cgi?id=1726226

**Comment by RPigott** [comment]
> It would probably work.

Here's a revived patch w/ Carlos's NULL buffer approach, since for some reason just cherry-picking it caused conflicts.

```diff
diff --git a/gdk/wayland/gdkwindow-wayland.c b/gdk/wayland/gdkwindow-wayland.c
index 2d7c42bd7a..07f77eb596 100644
--- a/gdk/wayland/gdkwindow-wayland.c
+++ b/gdk/wayland/gdkwindow-wayland.c
@@ -3286,7 +3286,7 @@ unmap_popups_for_window (GdkWindow *window)
 }
 
 static void
-gdk_wayland_window_hide_surface (GdkWindow *window)
+gdk_wayland_window_clear_surface (GdkWindow *window)
 {
   GdkWaylandDisplay *display_wayland = GDK_WAYLAND_DISPLAY (gdk_window_get_display (window));
   GdkWindowImplWayland *impl = GDK_WINDOW_IMPL_WAYLAND (window->impl);
@@ -3416,7 +3416,10 @@ gdk_wayland_window_hide_surface (GdkWindow *window)
 static void
 gdk_wayland_window_hide (GdkWindow *window)
 {
-  gdk_wayland_window_hide_surface (window);
+  GdkWindowImplWayland *impl = GDK_WINDOW_IMPL_WAYLAND (window->impl);
+
+  wl_surface_attach (impl->display_server.wl_surface, NULL, 0, 0);
+  wl_surface_commit (impl->display_server.wl_surface);
   _gdk_window_clear_update_area (window);
 }
 
@@ -3430,7 +3433,7 @@ gdk_window_wayland_withdraw (GdkWindow *window)
 
       g_assert (!GDK_WINDOW_IS_MAPPED (window));
 
-      gdk_wayland_window_hide_surface (window);
+      gdk_wayland_window_clear_surface (window);
     }
 }
 
@@ -3751,7 +3754,7 @@ gdk_wayland_window_destroy (GdkWindow *window,
    */
   g_return_if_fail (!foreign_destroy);
 
-  gdk_wayland_window_hide_surface (window);
+  gdk_wayland_window_clear_surface (window);
 }
 
 static void
```

and with this I can no longer reproduce.

Feel free to test. I'm waiting for logs from #6147 folks to understand how this manifests on multi-output setups, cause I'm pretty sure they're the same.

**Comment by chris-morgan** [comment]
gtk!3898 fixes all the menus and popups for me, so I can finally progress past the 2021-07-19 build of Firefox. I’m glad to say that tooltips are included in the things that are fixed: tooltips used to be unreliable (best hope of getting them was to hover over the thing with a tooltip for a second, then move the cursor a few pixels), now with this patch they’re finally reliable. I’m glad to see that that the one major breakage has caused other preexisting minor breakage to be resolved. 🙂

**Comment by stransky** [comment]
I see the missing callback / missing popups on Mutter too so it's not a Sway issue only. 

https://gitlab.gnome.org/GNOME/gtk/-/issues/1986 is why we do all the silly hide/show things here so I'm going to investigate if and when we really needs that.

**Comment by GreyXor** [comment]
@RPigott @ojab 
This problem appears for me only on my external screen 2 * scaled.

I confirm that https://gitlab.gnome.org/GNOME/gtk/-/merge_requests/3898 resolves popup and slowness problems for me in Firefox.

**Comment by stransky** [comment]
@RPigott Roman, I think the missing callback is not a showstopper here - I think I can create a workaround in Firefox if Gdk provides correct info that the parent GdkWindow is hidden (unmaped) even when gdk_wayland_window_get_wl_surface() returns some wl_surface.

So do you know if Gdk reports the hidden GdkWindow as mapped/visible or not?

**Comment by stransky** [comment]
I see it now. Gdk refuses to map a popup but sends a map signal and event and even returns wl_surface for it.

**Comment by RPigott** [comment]
@stransky I think those hide/show calls are irrelevant. Its the hide/show internal to gdk that matters. I tried to explain in https://gitlab.gnome.org/GNOME/gtk/-/issues/4198.

Gdk doesn't refuse to map a popup, it tries to. It creates the positioner, assigns the popup role and commits the surface. Then it sends a map signal. It does this before setting any scaling. When Gdk sends the map event firefox grabs a reference to Gdk's wl_surface and makes it's own subsurface to play with.

Then, when sway is preparing to map the surface, it sends the surface enter event for an output with >1 scale. When gdk receives this, it surreptitiously destroys the popup surface behind your back and creates a new one. Firefox's subsurface is now inert and the surface lock cannot be obtained. If you check out the full log I posted, you'll see that even though I was quick gdk goes on to create/destroy the surface several dozen times before I close the popup and firefox finally destroys the subsurface.

**Comment by RPigott** [comment]
Also regarding your linked bug, xdg_popup has since gained a reposition request which I think might be relevant. But I don't think gtk3 knows about it.

**Comment by stransky** [comment]
@RPigott It's possible that you see https://bugzilla.mozilla.org/show_bug.cgi?id=1708709 which is Firefox bug - popup is repeatedly moved/positioned for scaled output during to Firefox internal rounding error. And the hide/show workaround it a reason why you see popup move/hide.

I'm not aware of any Gdk internal code which does such popup foolery. Anyway, I'm going to fix that on Firefox side this week so we can confirm it's Firefox but or not.

**Comment by RPigott** [comment]
> It's possible that you see https://bugzilla.mozilla.org/show_bug.cgi?id=1708709 which is Firefox bug

No, that's not it. Sway sends the surface enter event before any configure, and since gdk destroys the popup surface immediately, sway never sends the popup a configure and gdk won't send the move-to-rect signal. Either way the subsurface resource is leaked in firefox – there's no reason for FF to hold onto that subsurface long after it's parent has been destroyed by gdk.

As I pointed out originally, this is the hide call responsible: https://gitlab.gnome.org/GNOME/gtk/-/blob/d4e2d05cd9518ba04d6fbe1cbcec27142788ac95/gdk/wayland/gdkwindow-wayland.c#L1227

**Comment by emersion** [comment]
> So may be the way here to report higher scales first? (I don't know how mutter handles this scenario.)

In general we don't add workarounds for buggy clients in our compositors. If there is a bug in a client, the client should be fixed.

**Comment by RPigott** [comment]
> I see that window hide/show sequence happens only when higher scale is reported after lower ones.

I'm not sure I see that... for the log I pasted above, I captured that log on my laptop with only a single display. There are two enter events because the client has bound two wl_output globals that refer to the same output. I'm not entirely sure why that is, but I assume maybe both gdk and firefox bind one somewhere. Obviously they have the same scale, first 1 and then 2 when I change it half way through.

However, there are definitely some extra shenanigans with multiple outputs that I'm still trying to understand in #6147. For starters, gdk often sends configure events with scale 1 and then scale 2 even when we send only one enter event, which sounds similar to what you describe.

I have some vague hypotheses about how the order of events might be messing with sway more than other compositors, but to investigate I need logs. If anyone wants to help out please visit #6147.

**Comment by stransky** [comment]
I just tested mutter with two displays and scales are send in scale 1 scale 2 order.

**Comment by stransky** [comment]
Yes, Firefox has it's own wl_output object - it's used to workaround mutter bug, mutter reports wrong resolutions for scaled displays. One wl_output object is owned by Gdk and it's used to set scale factor for GdkWindows.

**Comment by RPigott** [comment]
The only time a surface should receive enter events for two outputs is when it is displayed on both. This bug occurs even when only one output is connected. In my example, every enter event sent is for the same physical output which has scale=2. There is no scale=1 output. After only one enter event (on the wl_output bound by gdk) gdk destroys the surface.

The same is true for #6147, which occurs when the client surface is on only one monitor and receives only one enter event. It has nothing to do with the order of output scale events; those are sent when the client binds the output or the output scale changes.

Do you object to https://gitlab.gnome.org/GNOME/gtk/-/merge_requests/3898 ? That is my proposed solution.

**Comment by stransky** [comment]
> Do you object to https://gitlab.gnome.org/GNOME/gtk/-/merge_requests/3898 ? That is my proposed solution.

It looks fine to me.

**Comment by Samamy** [comment]
Hi all ! I think I found a temporary solution.
1) Disable scale for output or set `1` in `~/.config/sway/config`
2) Try firefox run like `GDK_DPI_SCALE=1.5 firefox`

**Comment by dngray** [comment]
> Try firefox run like `GDK_DPI_SCALE=1.5 firefox`

Can confirm that does fix the issue.

I've done a bit more [research on this](https://bugzilla.mozilla.org/show_bug.cgi?id=1730073).

Seems introduced in this commit https://hg.mozilla.org/integration/autoland/pushloghtml?fromchange=0c51eaf636b089097500fc721b31d99dcd92b0a3&tochange=02514549f3a78a80b922dc2224d039c6db2a10d8

**Comment by chris-morgan** [comment]
@dngray It has indeed been well established that that patch is what triggered this specific manifestation of the problem. Also everyone that has tried it here is agreeing that gtk!3898 fixes this problem and other less-obvious manifestations like tooltips which were mostly broken long before that change from mid-July.

**Comment by stransky** [comment]
I wonder how the commit can broke it - we just use correct screen sizes with the patch. 
Can anyone attach an output of firefor screen log:

MOZ_LOG="WidgetScreen:5" MOZ_ENABLE_WAYLAND=1 ./firefox

before and after the patch (i.e. Firefox 91 and 92) on Sway? I'd like to see what's different here.

**Comment by dngray** [comment]
> @dngray It has indeed been well established that that patch is what triggered this specific manifestation of the problem.

That may be the case. I'm just saying I noticed it today with the Firefox v92 update on Sway that wash pushed to me today. I did my own investigation independently because I didn't know about this issue or the other discussion about it.

**Comment by adia** [comment]
Here are two logs as asked for, under sway with scale 1.5:
[firefox-91.log](https://github.com/swaywm/sway/files/7141694/firefox-91.log)
[firefox-92.log](https://github.com/swaywm/sway/files/7141693/firefox-92.log)

**Comment by stransky** [comment]
Thanks. Which screen values do you get with Firefox 91 on this page? https://crisal.io/tmp/screen.html

**Comment by adia** [comment]
With Firefox 91:

    innerWidth: 1702
    innerHeight: 786
    outerWidth: 1702
    outerHeight: 903
    screen.width: 853
    screen.height: 480

For reference, with Nightly, the different values are:

    screen.width: 2560
    screen.height: 1440

**Comment by stransky** [comment]
Thanks, the 92 values looks correct and 91 wrong. I wonder how screen sizes can affects popups/sway by this way. https://bugzilla.mozilla.org/show_bug.cgi?id=1727073 also adds a fix for partial scales - this really puzzles me.

https://github.com/swaywm/sway/issues/6426#issuecomment-908777496  should not be the case here as there's only one monitor with one wl_output - so Sway should not send enter to wl_surface twice.

**Comment by RPigott** [comment]
Sway only sends one enter event per output global that refers to the entered output. For gdk in this context, output globals it did not bind behave as if they have scale 1.

Once again, you can see this in the log I posted. `wl_output@17` is bound by gdk, and `wl_output@31` is the new global bound by firefox, but they both refer to the same physical output. I added the "hide surface" log message to gdk to better capture the bad behavior, and annotated this excerpt with my explanation:

<details><summary>log excerpt</summary>

```
[...]
[882711.796] wl_surface@120.enter(wl_output@31) # scale 2, gdk thinks scale 1
Gdk-Message: 12:53:29.589: surface enter, window 0x7facf6917640 output 0x7fad19cd1150 # scale 1 -> 1, nothing
[882711.885] wl_surface@120.enter(wl_output@17) # scale 2
Gdk-Message: 12:53:29.589: surface enter, window 0x7facf6917640 output 0x7fad2c6f5830 # scale 1 -> 2, hide
Gdk-Message: 12:53:29.589: hide surface, window 0x7facf6917640
[...]
[882713.797] wl_surface@82.enter(wl_output@31) # scale 2, gdk thinks scale 1
Gdk-Message: 12:53:29.591: surface enter, window 0x7facf6917640 output 0x7fad19cd1150 # scale 2 -> 1, hide
Gdk-Message: 12:53:29.591: hide surface, window 0x7facf6917640
[...]
[882848.229] wl_surface@91.enter(wl_output@31) # scale 2, gdk thinks scale 1
Gdk-Message: 12:53:29.725: surface enter, window 0x7facf6917640 output 0x7fad19cd1150 # scale 1 -> 1, nothing
[882848.260] wl_surface@91.enter(wl_output@17) # scale 2
Gdk-Message: 12:53:29.725: surface enter, window 0x7facf6917640 output 0x7fad2c6f5830 # scale 1 -> 2, hide
Gdk-Message: 12:53:29.725: hide surface, window 0x7facf6917640
[...]
[882850.636] wl_surface@116.enter(wl_output@31) # scale 2, gdk thinks scale 1
Gdk-Message: 12:53:29.727: surface enter, window 0x7facf6917640 output 0x7fad19cd1150 # scale 2 -> 1, hide
Gdk-Message: 12:53:29.727: hide surface, window 0x7facf6917640
[...]
```
</details>

Your patch introduces the bug by adding a new output global that confuses gdk.

> https://bugzilla.mozilla.org/show_bug.cgi?id=1727073 also adds a fix for partial scales - this really puzzles me.

This whole business of trying to detect fractional scales is the wrong fix. Here you noticed the real problem: https://bugzilla.mozilla.org/show_bug.cgi?id=1727073#c6

> From some reason the popup is initially opened as small [...]
> I wonder if we miss a scale factor correction somewhere

In fact, I would bet that if you checked, KDE and Gnome just aren't sending you _any_ enter events before the popup is mapped. So you have no way to know what scale to draw the popup beforehand. I recommend you ask them to fix this because I consider it a bug.

Sway sends you an enter event _before_ the popup is mapped to prevent this: https://github.com/swaywm/sway/issues/6147#issuecomment-912857546.

But gdk doesn't expect that and does the hide/show shenanigans: https://gitlab.gnome.org/GNOME/gtk/-/issues/4198#note_1264658.

#6479 implements the "no-enter-before-map" bug in sway. I'm leaving it as a draft until the discussion on the wayland gitlab is resolved, at which point it will hopefully be unnecessary.

**Comment by RPigott** [comment]
@stranksy since gtk!3898 is rejected, I created https://gitlab.gnome.org/GNOME/gtk/-/merge_requests/3941 to fix the gdk bug with handling firefox's duplicate wl_outputs so that it doesn't loop. That should make it possible in principle to correct this behavior by doing as jadahl suggests.

I'd still like gdk to be capable of handling pre-map enter events correctly, but it will require a different approach.

**Comment by abraxxa** [comment]
This broke my Ubuntu + Sway setup. Do you have a workaround other than forcing XWayland by starting Firefox with `env -u WAYLAND_DISPLAY firefox`?
Version 92 seems to not need `MOZ_ENABLE_WAYLAND=1` any more, setting it to 0 also uses Wayland.

**Comment by toastal** [comment]
I moved to ESR (v91) til this fire settles. I've been on 91 since it was beta.
-- 
toastal ไข่ดาว | https://toast.al
7944 74b7 d236 dab9 c9ef
e7f9 5cce 6f14 66d4 7c9e

**Comment by stransky** [comment]
> This whole business of trying to detect fractional scales is the wrong fix. Here you noticed the real problem: https://bugzilla.mozilla.org/show_bug.cgi?id=1727073#c6

You're right but I don't think there is any any other way. Mutter reports screen sizes differently than KDE/Sway and I need wl_output to find correct screen size. Gtk does not export gdk_wayland_window_get_wl_output() so I need to get that from Wayland. It's definitely mutter/gtk weirdness here but I don't have any better solution now.

**Comment by rmader** [comment]
> You're right but I don't think there is any any other way. Mutter reports screen sizes differently than KDE/Sway and I need wl_output to find correct screen size. Gtk does not export gdk_wayland_window_get_wl_output() so I need to get that from Wayland. It's definitely mutter/gtk weirdness here but I don't have any better solution now.

Yeah, this one is unfortunate. AFAIK `wl_output` is simply not well defined for fractional scaling and if somehow possible we should only rely on `xdg-output`.

**Comment by stransky** [comment]
I think the best solution is to use wl_output workaround on Gnome only as it's needed for Mutter only.

**Comment by stransky** [comment]
https://bugzilla.mozilla.org/show_bug.cgi?id=1730476 should fix that.

**Comment by RPigott** [comment]
Resolved by https://gitlab.gnome.org/GNOME/gtk/-/merge_requests/3941.

**Comment by stransky** [comment]
Firefox 93 will fix that for old Gtk versions.

**Comment by SteadBytes** [comment]
The fix @stransky mentioned is currently available in [Firefox 93 Beta edition](https://www.mozilla.org/en-US/firefox/93.0beta/releasenotes/) for use prior to the Firefox 93 release. I've _personally_ had no issues running the beta but, of course, YMMV.

**Comment by abraxxa** [comment]
Thanks for the update! Where can such detailed bugfixes be found as it isn't mentioned in the release notes?

**Comment by SteadBytes** [comment]
See the [issue tracker link](https://github.com/swaywm/sway/issues/6426#issuecomment-918043190) posted previously by @stransky -  most specifically these two replies confirming the inclusion of the patch in 93 beta:
- https://bugzilla.mozilla.org/show_bug.cgi?id=1730476#c7
- https://bugzilla.mozilla.org/show_bug.cgi?id=1730476#c8

**Comment by abraxxa** [comment]
Why is this bug id marked as `status-firefox92: --- → unaffected`? Firefox 92 on Ubuntu is affected by this.

**Comment by SteadBytes** [comment]
Best I can do (I'm no expert either) is point towards the Mozilla Bugzilla search [help](https://bugzilla.mozilla.org/page.cgi?id=quicksearch.html) as well as the [advanced search](https://bugzilla.mozilla.org/query.cgi?format=advanced). There are options there for specifying a version in a search. 

Other than that, I suggest asking further questions on Mozilla Bugzilla and/or any related help channels as it is off topic for this thread :+1:

**Comment by abraxxa** [comment]
I can confirm that the issue is fixed by the Firefox 93 upgrade on Ubuntu 21.04.

**Comment by itslychee** [comment]
is this related to my issue with bitwarden's folder selection breaking the parent popup?

demo:
https://user-images.githubusercontent.com/82718618/189556171-bc2c2961-76aa-4f65-946a-1724397cc89c.mp4
